### ISSUE TITLE
Fails to load web3 on build

### ISSUE THREAD

**Issue by vanruch** [paritytech/parity-bridge#178]
When I run
`cargo build -p parity-bridge --release`
I get the following error:
```
error: failed to load source for a dependency on `web3`                         

Caused by:
  Unable to update https://github.com/tomusdrw/rust-web3?branch=bridge#aad14c09
```
And indeed, the branch https://github.com/tomusdrw/rust-web3/tree/bridge has been [deleted](https://github.com/tomusdrw/rust-web3/pull/129#event-1665465945)
When trying to install web3 from master, I get lots of errors, like for example
```
error[E0599]: no method named `logs` found for type `web3::api::Eth<T>` in the current scope
  --> bridge/src/api.rs:44:42
   |
44 |         future: api::Eth::new(transport).logs(filter),
   |                                          ^^^^
```

**Comment by snd** [comment]
someone deleted the `bridge` branch from the `rust-web3` repository since it was superseded by another branch.
for the moment you can just manually change the web3 dependency to:

```
web3 = { git = "https://github.com/forrest-marshall/rust-web3", branch = "get-logs" }
```

in all `Cargo.toml` of the bridge.

then it should build again.

this will be properly fixed on master once `snd-refactoring` is merged. will close this then

**Comment by marekkirejczyk** [comment]
Hi @snd, 

Thanks for the swift response. 
We have changed web3 dependency as described, got error pasted below:

```
$ cargo build -p parity-bridge --release
   Compiling bridge v0.4.0 (file:///Users/marekkirejczyk/Projects/ambrosus/parity-bridge/bridge)
   Compiling rustc_version v0.2.2
   Compiling rand v0.3.22
   Compiling parking_lot_core v0.2.13
   Compiling futures-cpupool v0.1.8
   Compiling aho-corasick v0.6.4
   Compiling heapsize v0.4.2
   Compiling unicase v2.1.0
   Compiling unicase v1.4.2
   Compiling crunchy v0.1.6
   Compiling syn v0.11.11
   Compiling idna v0.1.4
   Compiling error-chain v0.11.0
   Compiling thread_local v0.3.5
   Compiling kernel32-sys v0.2.2
   Compiling mio v0.6.13
   Compiling mime v0.2.6
   Compiling tokio-io v0.1.5
   Compiling parking_lot v0.5.4
   Compiling security-framework-sys v0.1.16
   Compiling core-foundation v0.2.3
   Compiling uint v0.1.2
   Compiling uint v0.2.1
   Compiling ethereum-types v0.2.3
   Compiling ethereum-types v0.3.2
   Compiling tiny-keccak v1.4.1
   Compiling fixed-hash v0.1.3
   Compiling fixed-hash v0.2.1
   Compiling regex v0.2.6
   Compiling remove_dir_all v0.3.0
   Compiling mio-uds v0.6.4
   Compiling tokio-core v0.1.12
   Compiling security-framework v0.1.16
   Compiling serde_derive_internals v0.19.0
   Compiling mime v0.3.5
   Compiling tempdir v0.3.6
   Compiling url v1.7.0
   Compiling tokio-proto v0.1.1
   Compiling tokio-uds v0.1.7
   Compiling env_logger v0.4.3
   Compiling serde_derive v1.0.27
   Compiling native-tls v0.1.5
   Compiling hyper v0.11.21
   Compiling hyper v0.10.13
   Compiling tokio-tls v0.1.4
   Compiling ethbloom v0.4.2
   Compiling ethbloom v0.5.0
   Compiling websocket v0.20.2
   Compiling hyper-tls v0.1.3
   Compiling jsonrpc-core v8.0.1
   Compiling ethabi v5.1.0
   Compiling docopt v0.8.3
   Compiling ethabi-derive v5.0.5
   Compiling web3 v0.3.0 (https://github.com/forrest-marshall/rust-web3?branch=get-logs#9387d9c0)
error[E0308]: mismatched types
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:181:40
    |
181 |             Token::Address(data) => Ok(data),
    |                                        ^^^^ expected struct `ethereum_types::H160`, found struct `ethereum_types::hash::H160`
    |
    = note: expected type `ethereum_types::H160`
               found type `ethereum_types::hash::H160`
note: Perhaps two different versions of crate `ethereum_types` are being used?
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:181:40
    |
181 |             Token::Address(data) => Ok(data),
    |                                        ^^^^

error[E0308]: mismatched types
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:187:24
    |
187 |         Token::Address(self)
    |                        ^^^^ expected struct `ethereum_types::hash::H160`, found struct `ethereum_types::H160`
    |
    = note: expected type `ethereum_types::hash::H160`
               found type `ethereum_types::H160`
note: Perhaps two different versions of crate `ethereum_types` are being used?
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:187:24
    |
187 |         Token::Address(self)
    |                        ^^^^

error[E0277]: the trait bound `ethereum_types::U256: std::convert::From<ethereum_types::uint::U256>` is not satisfied
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:196:59
    |
196 |           Token::Int(data) | Token::Uint(data) => Ok(data.into()),
    |                                                           ^^^^ the trait `std::convert::From<ethereum_types::uint::U256>` is not implemented for `ethereum_types::U256`
...
208 | uint_tokenizable!(U256, "U256");
    | -------------------------------- in this macro invocation
    |
    = help: the following implementations were found:
              <ethereum_types::U256 as std::convert::From<u16>>
              <ethereum_types::U256 as std::convert::From<&'a ethereum_types::U512>>
              <ethereum_types::U256 as std::convert::From<&'a [u8; 32]>>
              <ethereum_types::U256 as std::convert::From<ethereum_types::U512>>
            and 15 others
    = note: required because of the requirements on the impl of `std::convert::Into<ethereum_types::U256>` for `ethereum_types::uint::U256`

error[E0277]: the trait bound `ethereum_types::U128: std::convert::From<ethereum_types::uint::U256>` is not satisfied
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:196:59
    |
196 |           Token::Int(data) | Token::Uint(data) => Ok(data.into()),
    |                                                           ^^^^ the trait `std::convert::From<ethereum_types::uint::U256>` is not implemented for `ethereum_types::U128`
...
209 | uint_tokenizable!(U128, "U128");
    | -------------------------------- in this macro invocation
    |
    = help: the following implementations were found:
              <ethereum_types::U128 as std::convert::From<[u8; 16]>>
              <ethereum_types::U128 as std::convert::From<isize>>
              <ethereum_types::U128 as std::convert::From<i16>>
              <ethereum_types::U128 as std::convert::From<i8>>
            and 14 others
    = note: required because of the requirements on the impl of `std::convert::Into<ethereum_types::U128>` for `ethereum_types::uint::U256`

error[E0277]: the trait bound `ethereum_types::uint::U256: std::convert::From<ethereum_types::U256>` is not satisfied
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:202:26
    |
202 |         Token::Uint(self.into())
    |                          ^^^^ the trait `std::convert::From<ethereum_types::U256>` is not implemented for `ethereum_types::uint::U256`
...
208 | uint_tokenizable!(U256, "U256");
    | -------------------------------- in this macro invocation
    |
    = help: the following implementations were found:
              <ethereum_types::uint::U256 as std::convert::From<&'a [u8]>>
              <ethereum_types::uint::U256 as std::convert::From<i16>>
              <ethereum_types::uint::U256 as std::convert::From<ethereum_types::uint::U128>>
              <ethereum_types::uint::U256 as std::convert::From<[u8; 32]>>
            and 15 others
    = note: required because of the requirements on the impl of `std::convert::Into<ethereum_types::uint::U256>` for `ethereum_types::U256`

error[E0277]: the trait bound `ethereum_types::uint::U256: std::convert::From<ethereum_types::U128>` is not satisfied
   --> /Users/marekkirejczyk/.cargo/git/checkouts/rust-web3-55a1c3de6e9f5931/9387d9c/src/contract/tokens.rs:202:26
    |
202 |         Token::Uint(self.into())
    |                          ^^^^ the trait `std::convert::From<ethereum_types::U128>` is not implemented for `ethereum_types::uint::U256`
...
209 | uint_tokenizable!(U128, "U128");
    | -------------------------------- in this macro invocation
    |
    = help: the following implementations were found:
              <ethereum_types::uint::U256 as std::convert::From<&'a [u8]>>
              <ethereum_types::uint::U256 as std::convert::From<i16>>
              <ethereum_types::uint::U256 as std::convert::From<ethereum_types::uint::U128>>
              <ethereum_types::uint::U256 as std::convert::From<[u8; 32]>>
            and 15 others
    = note: required because of the requirements on the impl of `std::convert::Into<ethereum_types::uint::U256>` for `ethereum_types::U128`

error: aborting due to 6 previous errors

Some errors occurred: E0277, E0308.
For more information about an error, try `rust
```

I tried with master branch of `rust-web3`, but with the same effect. 
Interestingly if you download the repo, it will compile.

**Comment by snd** [comment]
>  Interestingly if you download the repo, it will compile.

which repo?

is this an acceptable temporary solution for you?

this is because after switching the branch different versions of ethereum types are used and their types are incompatible.

will get resolved soon

**Comment by marekkirejczyk** [comment]
@snd 
The `get-logs` and `master` branches on `https://github.com/forrest-marshall/rust-web3` compile correctly.
But `parity-bridge` doeas not. 
We were not able to compile it so far.

**Comment by snd** [comment]
this will be fixed for real once the refactoring branch is merged.

for the time being you could try bumping `ethereum-types` to `0.3` everywhere in the bridge.

rust-web3 `master` and `get-logs` use `0.3` whereas the old `bridge` branch used `0.2` afaik

if that still doesn't work you might need to bump ethabi by one version. just post the build errors and i'll investigate

**Comment by vanruch** [comment]
Thanks, it worked after changing 
`pub fn logs<T: Transport>(transport: T, filter: &Filter)` to `pub fn logs<T: Transport>(transport: T, filter: Filter) `

**Comment by mills-nick** [comment]
I too wanted to evaluate the Parity Bridge, but due to this problem my efforts have stalled. Is it possible to push a buildable branch to this repo?

**Comment by marekkirejczyk** [comment]
It seems it might take some more time for the bridge to be stable.
You might want to check parity-bridge fork for now: https://github.com/poanetwork/poa-bridge

**Comment by mills-nick** [comment]
Thanks, marekkirejczyk. You inspired me to make my own fork from the latest paritytech master at: https://github.com/mills-nick/parity-bridge. This builds, now for the deploy...

**Comment by snd** [comment]
@mills-nick thanks for fixing the build! if you do a PR from your branch i'll merge it into master

**Comment by Starfunk** [comment]
@mills-nick I've managed to build the bridge from your fork, but I am now having trouble getting `parity-bridge` to read the `bridge-config.toml`. I've installed `parity-bridge` in my `$PATH`.  I created a new directory `/maximiliankahn/ParityBridge` where I put my `bridge_config.toml` and an empty `bridge.db` file.

 I then run `env RUST_LOG=info parity-bridge --config ~/ParityBridge/bridge_config.toml --database bridge.db`. The output is: 

```
INFO:parity_bridge: Parsing cli arguments
INFO:parity_bridge: Loading config
No such file or directory (os error 2)
```
Has anyone had any success deploying the parity-bridge? 

I also noticed that testing the build (i.e. running `cargo test`) results in the following errors:

```
error[E0277]: the trait bound `web3::types::Log: std::default::Default` is not satisfied
   --> bridge/src/bridge/deposit_relay.rs:163:15
    |
163 |             ..Default::default()
    |               ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `web3::types::Log`
    |
    = note: required by `std::default::Default::default`

error[E0277]: the trait bound `web3::types::Log: std::default::Default` is not satisfied
   --> bridge/src/bridge/withdraw_relay.rs:331:15
    |
331 |             ..Default::default()
    |               ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `web3::types::Log`
    |
    = note: required by `std::default::Default::default`

error[E0277]: the trait bound `web3::types::Log: std::default::Default` is not satisfied
   --> bridge/src/bridge/withdraw_relay.rs:365:15
    |
365 |             ..Default::default()
    |               ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `web3::types::Log`
    |
    = note: required by `std::default::Default::default`

error: aborting due to 3 previous errors
```

**Comment by jchen86** [comment]
@Starfunk the documentation is slightly outdated, you need to build and run the `parity-bridge-deploy` module to deploy.

and run
```
$ env RUST_LOG=info parity-bridge-deploy --config bridge_config.toml --database bridge_db.toml
```

I have a working setup & instructions here (deployment, making a deposit & withdraw)
https://github.com/fernnetwork/fern-parity-bridge-scripts

**Comment by Starfunk** [comment]
@jchen86  Thanks! Got it running now.
